Migrate to vitest from jest

mattzcarey opened the pull request:
<!-- Provide a brief summary of your changes -->
its newer, faster, works natively with the modern toolchain like typescript. 

## Motivation and Context
<!-- Why is this change needed? What problem does it solve? -->

## How Has This Been Tested?
<!-- Have you tested this in a real application? Which scenarios were tested? -->

## Breaking Changes
<!-- Will users need to update their code or configurations? -->

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [x] I have added or updated documentation as needed

## Additional context
<!-- Add any other context, implementation notes, or design decisions -->

felixweinberger commented:
Support shipping this as this seems like a pareto improvement. Looks like there are some conflicts to resolve.

Looking at CI: 

<img width="986" height="393" alt="CleanShot 2025-11-11 at 19 26 00" src="https://github.com/user-attachments/assets/2946fd76-60a9-47dc-b009-32e5628a6829" />

Not sure if these 2 are stalling because of this PR

mattzcarey committed:
wip

mattzcarey committed:
remove setup.js

mattzcarey committed:
remove msw

mattzcarey committed:
vitest types

mattzcarey committed:
more type fixes

mattzcarey committed:
unused packages

mattzcarey committed:
remove unneded timeouts

mattzcarey committed:
add lib mocking back

mattzcarey committed:
remove timeout

mattzcarey commented on src/server/streamableHttp.test.ts:
```diff
@@ -1369,16 +1376,19 @@ describe('StreamableHTTPServerTransport with resumability', () => {
     });
 
-    it('should store and replay MCP server tool notifications', async () => {
+    it('should store and replay MCP server tool notifications', { timeout: 10000 }, async () => {
```
removed

pcarleton commented on src/client/auth.test.ts:
```diff
@@ -1110,17 +1119,10 @@ describe('OAuth Authorization', () => {
             });
 
             expect(tokens).toEqual(validTokens);
-            expect(mockFetch).toHaveBeenCalledWith(
```
for my education, why was this necessary? does vitest not support this matcher pattern?

mattzcarey replied:
no it supports it fine my bad. Really a better solution here is to use toEqual() rather than contains. since contains can include a bunch of other stuff which might break the test case. eg. headers.

KKonstantinov replied:
Hey, this is true - you can't match it because it's a new instance of Headers (in vitest, this applies for any class instance).

But should you not be able to do:

```typescript
expect(mockFetch).toHaveBeenCalledWith(
                expect.objectContaining({
                    href: 'https://auth.example.com/token'
                }),
                expect.objectContaining({
                    method: 'POST',
                    headers: expect.objectContaining({
                        'Content-Type': 'application/x-www-form-urlencoded'
                    })
                })
            );
```

`expect(mockFetch.mock.calls[0][1])` would be flagged up by some linting rules (if ever added), because a [0] and a [1] element might not exist, so calling `.headers` on undefined is a possibility, where as toHaveBeenCalledWith is safe.

With vitest, we could achieve pretty much everything without having to resort to `.mock.calls` and similar - as these resort to testing actual implementation (e.g. in the test, we're hardcoding the exact order in which something gets called) as opposed to just testing the outcome.

pcarleton replied:
yea ideally we'd figure out how to do the matcher here, the [0][1] is a bit jank.  anyway, we can always fix later, let's move this on since it'll be a big merge conflict for other stuff if we leave it longer

pcarleton commented on src/server/streamableHttp.test.ts:
```diff
@@ -1288,9 +1288,16 @@ describe('StreamableHTTPServerTransport with resumability', () => {
-            // For test simplicity, just return all events with matching streamId that aren't the lastEventId
-            for (const [eventId, { message }] of storedEvents.entries()) {
-                if (eventId.startsWith(streamId) && eventId !== lastEventId) {
+            // Convert to array and find all events after the lastEventId
```
what happened here? is this related?

mattzcarey replied:
its a bit more readable. functionally the same. reverting.

pcarleton reviewed (changes requested):
overall looks good, ty for doing this. just had 2 questions about parts i didn't understand

mattzcarey committed:
pls vite plsss

felixweinberger committed:
Merge branch 'main' into migrate-to-vitest

felixweinberger committed:
Fix incomplete vitest migration in auth.test.ts

Replace remaining Jest-specific APIs with Vitest equivalents:
- jest.Mock -> Mock (already imported from vitest)
- jest.fn() -> vi.fn()
- Add non-null assertion for TypeScript strict null check

morozow commented:
I also noticed that Vitest is used across several repositories, and that there was a migration from Jest to Vitest. I left a more detailed comment in the server repository here: https://github.com/modelcontextprotocol/servers/pull/2884#issuecomment-4372281511
I recommend reconsidering this decision carefully, because there are potential stability and diagnostic concerns that should be reviewed before this pattern spreads further across repositories. The async/generator/mocked-function behavior mentioned there is only one small example of the broader concern. It should not be treated as the only reason. The main point is that Vitest can introduce different runtime and diagnostic behavior compared to Jest, and these differences may affect agents and developers who need stable, predictable test execution and clear failure signals.
For worldwide projects, including OSS projects, I would prefer to keep the test flow on Jest unless Vitest behavior is explicitly validated against these kinds of edge cases.

morozow commented:
This migration introduced `vite` as a mandatory peer dep of vitest. The currently locked `vite@7.3.0` has two high-severity CVEs (GHSA-v2wj-q39q-566r, GHSA-p9ff-h696-f583), patched in ≥7.3.2. `pnpm.overrides` does not force the resolution for peer deps, so `pnpm audit --audit-level=high` fails on a clean checkout with no workaround short of adding `vite` as an explicit devDependency. Jest didn't have this problem — no bundler in the dep tree.
